### PR TITLE
finish off psr-18 auto-instrumentation

### PR DESCRIPTION
**Pull request by brettmc**
- add tests
- replace enum with class for php8.0 compatibility
- set up auto-registration via composer autoloader

### CODE REVIEW

File: src/Instrumentation/Psr18/src/Psr18Instrumentation.php [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+            'sendRequest',
+            pre: static function (ClientInterface $client, array $params, string $class, string $function, ?string $filename, ?int $lineno): ?array {
+                $request = $params[0] ?? null;
+                if (!$request instanceof RequestInterface) {
+                    Context::storage()->attach(Context::getCurrent());
+
+                    return null;
+                }
+
+                static $instrumentation;
```
Comment (pdelewski) [inline src/Instrumentation/Psr18/src/Psr18Instrumentation.php:39]: Can we move this instance to `register` function and follow the same pattern as for PSR-15?

File: src/Instrumentation/Psr18/src/Psr18Instrumentation.php [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+
+                Context::storage()->attach($context);
+
+                return [$request];
+            },
+            post: static function (ClientInterface $client, array $params, ?ResponseInterface $response, ?Throwable $exception): void {
+                $scope = Context::storage()->scope();
+                $scope?->detach();
+
+                if (!$scope || $scope->context() === Context::getCurrent()) {
```
Comment (pdelewski) [inline src/Instrumentation/Psr18/src/Psr18Instrumentation.php:83]: As I mentioned during review of PSR-15, it's not clear for me why there is part after or operator `$scope->context() === Context::getCurrent()` ?

File: src/Instrumentation/Psr18/src/Psr18Instrumentation.php [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+                $span = Span::fromContext($scope->context());
+
+                if ($response) {
+                    $span->setAttribute(TraceAttributes::HTTP_STATUS_CODE, $response->getStatusCode());
+                    $span->setAttribute(TraceAttributes::HTTP_FLAVOR, $response->getProtocolVersion());
+                    $span->setAttribute(TraceAttributes::HTTP_RESPONSE_CONTENT_LENGTH, $response->getHeaderLine('Content-Length'));
+
+                    foreach ((array) (get_cfg_var('otel.instrumentation.http.response_headers') ?: []) as $header) {
+                        if ($response->hasHeader($header)) {
+                            $span->setAttribute(sprintf('http.response.header.%s', strtr(strtolower($header), ['-' => '_'])), $response->getHeader($header));
```
Comment (pdelewski) [inline src/Instrumentation/Psr18/src/Psr18Instrumentation.php:96]: nitpick: would not be cleaner extract `sprintf` into separate line?

File: src/Instrumentation/Psr18/README.md [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+## Overview
+Auto-instrumentation hooks are registered via composer, and spans will automatically be created for each PSR-15 middleware that is executed.
+
+To export spans, you will need to create and register a `TracerProvider` early in your application's lifecycle:
+
+```php
+<?php
+require_once 'vendor/autoload.php';
+
+$tracerProvider = /*create tracer provider*/;
```
Comment (pdelewski) [inline src/Instrumentation/Psr18/README.md:19]: Do we still need do this that way as https://github.com/open-telemetry/opentelemetry-php/pull/854 is already merged?

Comment (brettmc) [reply src/Instrumentation/Psr18/README.md:23]: I've updated the text so that sdk autoloading is preferred... how does it sound now?

Comment (pdelewski) [reply src/Instrumentation/Psr18/README.md:23]: Sounds good to me

**Review by pdelewski — APPROVED** [review]

**Pull request merged**